File: app/src/main/java/eu/darken/bluemusic/main/core/service/modules/events/AppLaunchModule.java
package eu.darken.bluemusic.main.core.service.modules.events;

import javax.inject.Inject;

import eu.darken.bluemusic.AppComponent;
import eu.darken.bluemusic.bluetooth.core.SourceDevice;
import eu.darken.bluemusic.data.device.ManagedDevice;
import eu.darken.bluemusic.main.core.service.modules.EventModule;
import eu.darken.bluemusic.util.AppTool;
import timber.log.Timber;

@AppComponent.Scope
public class AppLaunchModule extends EventModule {
    private final AppTool appTool;

    @Inject
    AppLaunchModule(AppTool appTool) {
        super();
        this.appTool = appTool;
    }

    @Override
    public int getPriority() {
        return 1;
    }

    @Override
    public void handle(ManagedDevice eventDevice, SourceDevice.Event event) {
        if (event.getType() != SourceDevice.Event.Type.CONNECTED) return;
        if (eventDevice.getLaunchPkg() == null) return;

        Timber.d("Launching app %s", eventDevice.getLaunchPkg());
        appTool.launch(eventDevice.getLaunchPkg());

        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Timber.w(e);
        }
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/core/service/modules/events/AppLaunchModule.kt
package eu.darken.bluemusic.main.core.service.modules.events

import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoSet
import eu.darken.bluemusic.bluetooth.core.SourceDevice
import eu.darken.bluemusic.common.AppTool
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.WARN
import eu.darken.bluemusic.common.debug.logging.asLog
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.devices.core.ManagedDevice
import eu.darken.bluemusic.main.core.service.modules.EventModule
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AppLaunchModule @Inject constructor(
    private val appTool: AppTool
) : EventModule {

    override val priority: Int = 1

    override suspend fun handle(device: ManagedDevice, event: SourceDevice.Event) {
        if (event.type != SourceDevice.Event.Type.CONNECTED) return
        if (device.launchPkg == null) return

        log(TAG) { "Launching app ${device.launchPkg}" }
        appTool.launch(device.launchPkg!!)

        try {
            Thread.sleep(1000)
        } catch (e: InterruptedException) {
            log(TAG, WARN) { "Thread interrupted: ${e.asLog()}" }
        }
    }

    @Module @InstallIn(SingletonComponent::class)
    abstract class Mod {
        @Binds @IntoSet abstract fun bind(entry: AppLaunchModule): EventModule
    }

    companion object {
        private val TAG = logTag("AppLaunchModule")
    }
}